app/src/main/java/com/topjohnwu/magisk/utils/RootUtils.kt

```diff
@@ -158,7 +158,21 @@ class RootUtils : Shell.Initializer() {
 
         @JvmStatic
         fun reboot() {
-            Shell.su("/system/bin/reboot ${if (Config.recovery) "recovery" else ""}").submit()
+            reboot(if (Config.recovery) "recovery" else "")
+        }
+
+        @JvmStatic
+        fun reboot(reason: String) {
+            Shell.su("service call power ${getRebootTransactionCode()} i32 0 s16 $reason i32 0").submit()
+        } 
+
+        private fun getRebootTransactionCode(): Int {
+            // Get the transaction code for the reboot method in PowerManager.
+            // Warning: Hidden-api usage.
+            val clz = java.lang.Class.forName("android.os.IPowerManager\$Stub")
+            val field = clz.getDeclaredField("TRANSACTION_reboot")
```

Comment:
Don't assume the field of `reboot` transaction code is `TRANSACTION_reboot`, the format may be `TRANSACTION_reboot_XX` (XX is number).

Reply:
Could you give me an example of how that XX works? Thanks.

Reply:
I only know some examples, such as `TRANSACTION_getInstalledApplications_49 = 50` appears on API 25.

Reply:
> I only know some examples, such as `TRANSACTION_getInstalledApplications_49 = 50` appears on API 25.

Thanks.